### PR TITLE
Compress HTML

### PR DESCRIPTION
**Pull request by doktorbro**
The layout compresses HTML without plugins. It’s well [documented](http://jch.penibelst.de/).

### CODE REVIEW

File: _blog.config.yml [hunk 1]
```diff
@@ -17,14 +17,14 @@ avatar: /images/blog-logo.svg
 
 # Includes an icon in the footer for each username you enter
 footer-links:
-  dribbble: 
-  email: 
-  facebook: 
-  flickr:
```
Comment (doktorbro) [inline _blog.config.yml:23]: This whitespace is deleted by my editor. Sorry.

Comment (jc4p) [inline _blog.config.yml:23]: Nothing wrong with cleaning up unneeded whitespace :)

**Comment by NickCraver** [comment]
As a heads up: this will be reverted by the overhaul in paging I'm doing now. The server we're on are gzipping the result as well as we're caching at the edge delivering the HTML very fast. The blunt honest analysis here is the compression adds very little, but has several downsides:
- It's harder to work on the site
- Build time is increased
- The space savings above and beyond gzip are almost always sub-single-packet size

I don't meant to argue against the work done here - all contributions are very much appreciated. As a courtesy though, I wanted to give you a heads up my incoming PR will be removing this layout it.

**Comment by doktorbro** [comment]
@NickCraver Thank you for analysing. Maybe we have a little misunderstanding here. Let me clear it.

> It's harder to work on the site

My layout provides the option `ignore`. There you can set your development environment, so the layout is _disabled_ during you work on the site.

> Build time is increased

How much in percent? For now you have some options set to `all`. If you pick only the elements you use on the site, the layout will do less and the build time will decrease.

> The space savings above and beyond gzip are almost always sub-single-packet size

How many bytes on the home page? How many bytes on the post _[Going From Mobile Back To The Web](http://blog.stackexchange.com/2015/07/going-from-mobile-back-to-the-web/)_?

The option `profile` shows the page size after every compression step. If you are interested, you can see explicit numbers.

**Comment by NickCraver** [comment]
Hey @penibelst, I don't think we have a misunderstanding - just a disagreement of benefits :) The "it's harder to work on the site" stays true. Either I can have readable source (both locally _and_ in production), or I can work easier locally but lose confidence since I'm not actually testing the same thing as I'm deploying.

The build time isn't a huge factor, what we want is the fastest experience for the end user. That being said, you have to set `all` for it to have a fighting chance there, else you're just wasting time and speed. With that in mind, I've just setup tests of the new site (which is all server-side generation in PR #118), not the very tiny HTML + large JSON load delay we had before. This test is on my local laptop using the same gzip configuration we have on the production web tier in IIS.

Here's what the situation looks like:
- /2015/07/going-from-mobile-back-to-the-web/ (Blog Post)
  - Raw: 20,387
  - Raw Gzip: 8,999
  - Jekyll Compress: 17,667
  - Jekyll Compress Gzip: 8,497
  - Savings: **502 bytes**
- / (Homepage)
  - Raw: 15,578
  - Raw Gzip: 4,352
  - Jekyll Compress: 13,147
  - Jekyll Compress Gzip: 4,824
  - Savings:  **-472 bytes**
- Generation
  - Raw 50: done in 5.704 seconds.
  - Raw All: done in 25.058 seconds.
  - Compressed 50: done in 7.378 seconds. (29% slower)
  - Compressed All: done in 44.094 seconds. (76% slower)

On the home page (due to content being repeated), stripping elements down and making repeated chunks less-repeated actually has an adverse impact on overall zipped page size. On the blog post there's a slight savings, but it's small enough to make zero difference (literally zero, not _effectively_ zero) in transmission time to the client. Those are the numbers you can measure - which show no actual gains here aside from disk space usage. The build time is obviously significantly slower, but not prohibitively so. It's doing more work and taking more time - this is an common tradeoff I won't dig into here.

Here's the less measurable benefit of not compressing at the HTML level: which of these is easier to work on?

Uncompressed

Compressed

Not only for working on, it's easier to browse, etc. Our main audience is developers, so the code being pretty is a plus...but also not an item I'm weighing on usage here.

I recommend further reading on the [TCP congestion window](https://en.wikipedia.org/wiki/Congestion_window) (of which Stack Exchange uses 10) that makes the actual impact to users zero if you're curious (you seem to be!). The simple/harsh fact here is that so few of our pages hit that with gzip that this _literally_ doesn't matter. After author pagination only a very, _very_ lengthy blog post (way longer than any existing) would actually make the bar here.

**Comment by doktorbro** [comment]
@NickCraver Thank you for the numbers. I will rethink some ideas.

Last question: what means “50” in the generation statistics?

**Comment by NickCraver** [comment]
Hey @penibelst,
I tested both generating only the latest 50 posts (`limit_posts: 50` in config) vs. commenting it out and generating all posts.  I did this because the local dev config is to generate 50 to be as fast at revving as possible and the size tests were done generating all since so it's a level playing field all around.

**Pull request merged**